File: backend/src/test/java/de/iteratec/konfi/KonfiTest.java
package de.iteratec.konfi;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;

import javax.transaction.Transactional;

import static org.hamcrest.Matchers.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class KonfiTest {

    @Autowired
    private MockMvc mvc;

    @Test
    public void createAndList() throws Exception {
        JSONObject json = createConferenceJson();

        postJson("/conference", json)
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id", greaterThan(0)));

        mvc.perform(get("/conference"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", hasSize(1)))
                .andExpect(jsonPath("$[0].name", is(json.getString("name"))))
                .andExpect(jsonPath("$[0].maxAttendees", is(json.getInt("maxAttendees"))))
                .andExpect(jsonPath("$[0].startDate", is(json.getString("startDate"))))
                .andExpect(jsonPath("$[0].endDate", is(json.getString("endDate"))));
    }

    @Test
    public void createInvalid() throws Exception {
        JSONObject json = createConferenceJson();
        postJson("/conference", json.put("name", null)).andExpect(status().isBadRequest());
        postJson("/conference", json.put("name", "a")).andExpect(status().isBadRequest());
        postJson("/conference", json.put("name", "a".repeat(60))).andExpect(status().isBadRequest());
        postJson("/conference", json.put("maxAttendees", null)).andExpect(status().isBadRequest());
        postJson("/conference", json.put("maxAttendees", -1)).andExpect(status().isBadRequest());
        postJson("/conference", json.put("startDate", null)).andExpect(status().isBadRequest());
        postJson("/conference", json.put("startDate", "1970-01-03T09:00:00")).andExpect(status().isBadRequest());
        postJson("/conference", json.put("endDate", null)).andExpect(status().isBadRequest());
    }

    @Test
    public void createColliding() throws Exception {
        JSONObject json = createConferenceJson();
        postJson("/conference", json).andExpect(status().isOk());
        postJson("/conference", json).andExpect(status().isBadRequest());
    }

    @Test
    public void register() throws Exception {
        Long conferenceId = createConference(createConferenceJson());
        JSONObject attendeeJson = createAttendeeJson();
        MockHttpServletRequestBuilder registerRequest = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(attendeeJson.toString());

        mvc.perform(registerRequest).andExpect(status().isOk());
        mvc.perform(get("/conference/{id}/attendees", conferenceId))
                .andExpect(jsonPath("$", hasSize(1)))
                .andExpect(jsonPath("$[0].email", is(attendeeJson.getString("email"))))
                .andExpect(jsonPath("$[0].firstName", is(attendeeJson.getString("firstName"))))
                .andExpect(jsonPath("$[0].lastName", is(attendeeJson.getString("lastName"))));
    }

    @Test
    public void registerInvalidAttendee() throws Exception {
        Long conferenceId = createConference(createConferenceJson());
        JSONObject attendeeJson = createAttendeeJson().put("email", "a");
        MockHttpServletRequestBuilder registerRequest = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(attendeeJson.toString());

        mvc.perform(registerRequest).andExpect(status().isBadRequest());
    }

    @Test
    public void registerDuplicateEmail() throws Exception {
        Long conferenceId = createConference(createConferenceJson());
        JSONObject attendeeJson = createAttendeeJson();
        MockHttpServletRequestBuilder registerRequest = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(attendeeJson.toString());

        mvc.perform(registerRequest).andExpect(status().isOk());
        mvc.perform(registerRequest).andExpect(status().isBadRequest());
    }

    @Test
    public void registerMaxAttendeesExceeded() throws Exception {
        Long conferenceId = createConference(createConferenceJson().put("maxAttendees", 1));
        MockHttpServletRequestBuilder registerRequest = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(createAttendeeJson().put("email", "test1@konfi.io").toString());
        mvc.perform(registerRequest).andExpect(status().isOk());

        MockHttpServletRequestBuilder registerRequest2 = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(createAttendeeJson().put("email", "test2@konfi.io").toString());
        mvc.perform(registerRequest2).andExpect(status().isBadRequest());
    }

    @Test
    public void deregister() throws Exception {
        Long conferenceId = createConference(createConferenceJson());
        JSONObject attendeeJson = createAttendeeJson();
        MockHttpServletRequestBuilder registerRequest = post("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(attendeeJson.toString());
        MockHttpServletRequestBuilder deregisterRequest = delete("/conference/{id}/attendees", conferenceId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(attendeeJson.toString());

        mvc.perform(registerRequest).andExpect(status().isOk());
        mvc.perform(deregisterRequest).andExpect(status().isOk());
        mvc.perform(get("/conference/{id}/attendees", conferenceId))
                .andExpect(jsonPath("$", hasSize(0)));
    }

    private Long createConference(JSONObject conferenceJson) throws Exception {
        MockHttpServletRequestBuilder request = post("/conference")
                .contentType(MediaType.APPLICATION_JSON)
                .content(conferenceJson.toString());

        String createdConferenceJson = mvc.perform(request).andReturn().getResponse().getContentAsString();
        return new JSONObject(createdConferenceJson).getLong("id");
    }

    private ResultActions postJson(String path, JSONObject json) throws Exception {
        return mvc.perform(post(path).contentType(MediaType.APPLICATION_JSON).content(json.toString()));
    }

    private JSONObject createConferenceJson() throws Exception {
        return new JSONObject()
                .put("name", "KotlinKonf")
                .put("maxAttendees", 200)
                .put("startDate", "2021-01-03T09:00:00")
                .put("endDate", "2021-01-05T18:00:00");
    }

    private JSONObject createAttendeeJson() throws Exception {
        return new JSONObject()
                .put("firstName", "Max")
                .put("lastName", "Mustermann")
                .put("email", "maxi@konfi.io");
    }
}

File: backend/src/test/kotlin/de/iteratec/konfi/KonfiTest.kt
package de.iteratec.konfi

import org.hamcrest.Matchers.*
import org.json.JSONObject
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import javax.transaction.Transactional

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
class KonfiTest {

    @Autowired
    private lateinit var mvc: MockMvc

    @Test
    fun `created conference appears on a conference list`() {
        val json = createConferenceJson()

        post("/conference")
            .jsonContent(json)
            .execute()
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.id", greaterThan(0)))

        get("/conference")
            .execute()
            .andExpect(status().isOk)
            .andExpect(jsonPath("$", hasSize<Int>(1)))
            .andExpect(jsonPath("$[0].name", equalTo(json.getString("name"))))
            .andExpect(jsonPath("$[0].maxAttendees", equalTo(json.getInt("maxAttendees"))))
            .andExpect(jsonPath("$[0].startDate", equalTo(json.getString("startDate"))))
            .andExpect(jsonPath("$[0].endDate", equalTo(json.getString("endDate"))))
    }

    @Test
    fun `conferences with invalid attributes are refused`() {
        fun assertFails(json: JSONObject) {
            post("/conference").jsonContent(json).execute().andExpect(status().isBadRequest)
        }

        val json = createConferenceJson()
        assertFails(json.put("name", null))
        assertFails(json.put("name", "a"))
        assertFails(json.put("name", "a".repeat(60)))
        assertFails(json.put("maxAttendees", null))
        assertFails(json.put("maxAttendees", -1))
        assertFails(json.put("startDate", null))
        assertFails(json.put("startDate", "1970-01-03T09:00:00"))
        assertFails(json.put("endDate", null))
    }

    @Test
    fun `simultaneous conferences with the same name are not allowed`() {
        val json = createConferenceJson()
        post("/conference").jsonContent(json).execute().andExpect(status().isOk)
        post("/conference").jsonContent(json).execute().andExpect(status().isBadRequest)
    }

    @Test
    fun `registered attendee appears on the attendee list`() {
        val conferenceId = createConference(createConferenceJson())
        val attendeeJson = createAttendeeJson()

        post("/conference/{id}/attendees", conferenceId)
            .jsonContent(attendeeJson)
            .execute()
            .andExpect(status().isOk)

        get("/conference/{id}/attendees", conferenceId)
            .execute()
            .andExpect(jsonPath("$", hasSize<Int>(1)))
            .andExpect(jsonPath("$[0].email", equalTo(attendeeJson.getString("email"))))
            .andExpect(jsonPath("$[0].firstName", equalTo(attendeeJson.getString("firstName"))))
            .andExpect(jsonPath("$[0].lastName", equalTo(attendeeJson.getString("lastName"))))
    }

    @Test
    fun `attendees with invalid attributes are refused`() {
        val conferenceId = createConference(createConferenceJson())
        val attendeeJson = createAttendeeJson().put("email", "a")

        post("/conference/{id}/attendees", conferenceId)
            .jsonContent(attendeeJson)
            .execute()
            .andExpect(status().isBadRequest)
    }

    @Test
    fun `attendee may register for a conference only once`() {
        val conferenceId = createConference(createConferenceJson())
        val attendeeJson = createAttendeeJson()

        val registerRequest = post("/conference/{id}/attendees", conferenceId).jsonContent(attendeeJson)
        registerRequest.execute().andExpect(status().isOk)
        registerRequest.execute().andExpect(status().isBadRequest)
    }

    @Test
    fun `maximal attendee limit cannot be exceeded`() {
        val conferenceId = createConference(createConferenceJson().put("maxAttendees", 1))

        post("/conference/{id}/attendees", conferenceId)
            .jsonContent(createAttendeeJson().put("email", "test1@konfi.io"))
            .execute()
            .andExpect(status().isOk)

        post("/conference/{id}/attendees", conferenceId)
            .jsonContent(createAttendeeJson().put("email", "test2@konfi.io"))
            .execute()
            .andExpect(status().isBadRequest)
    }

    @Test
    fun `deregistered attendee disappears from the attendee list`() {
        val conferenceId = createConference(createConferenceJson())
        val attendeeJson = createAttendeeJson()

        post("/conference/{id}/attendees", conferenceId)
            .jsonContent(attendeeJson)
            .execute()
            .andExpect(status().isOk)

        delete("/conference/{id}/attendees", conferenceId)
            .jsonContent(attendeeJson)
            .execute()
            .andExpect(status().isOk)

        get("/conference/{id}/attendees", conferenceId)
            .execute()
            .andExpect(jsonPath("$", hasSize<Int>(0)))
    }

    private fun createConference(conferenceJson: JSONObject): Long {
        return post("/conference")
            .jsonContent(conferenceJson)
            .execute()
            .andReturn().response.contentAsString
            .let { JSONObject(it).getLong("id") }
    }

    private fun MockHttpServletRequestBuilder.jsonContent(json: JSONObject) = this
        .contentType(MediaType.APPLICATION_JSON)
        .content(json.toString())

    private fun MockHttpServletRequestBuilder.execute() = mvc.perform(this)

    private fun createConferenceJson() = JSONObject().apply {
        put("name", "KotlinKonf")
        put("maxAttendees", 200)
        put("startDate", "2021-01-03T09:00:00")
        put("endDate", "2021-01-05T18:00:00")
    }

    private fun createAttendeeJson() = JSONObject().apply {
        put("firstName", "Max")
        put("lastName", "Mustermann")
        put("email", "maxi@konfi.io")
    }
}